### DISCUSSION TITLE
how to run commands through the user shell with deno?

### DISCUSSION THREAD

**Discussion by joaomelo** [denoland/deno#12493] [Q&A]
I'm trying to build a  simple CLI tool to run dev scripts with Deno. The goal is to create cross-platform scripts with type safety.

I'm not succeeding with Deno.run(). I think this is due to part of my scripts are based on global packages installed with NPM like, for example, the firebase CLI tool.

If i run something like this:
```js
  const { code } = await Deno.run({
    cmd: ["firebase", "emulators:start"],
  }).status();
```

I get this error
```
error: Uncaught (in promise) NotFound: O sistema não pode encontrar o arquivo especificado. (os error 2)
  const { code } = await Deno.run({
```

I found some clarification in this [issue](https://github.com/denoland/deno/issues/5921) about echo. But, I'm afraid if I prefix everything with `"cmd", "/c",` the program will tie itself to the OS. Is this a valid assumption?

Is there another API that enables us to run commands through the current shell instead of attempt to find binaries in the system directly?

I think in node this is done with something like this:
``` js
    const p = spawn("firebase emulators:start", {
      shell: true,
    });
```

Thanks for the awesome tool. 
JM

**Comment by iugo** [comment]
demo:

```ts
// git commit
const res = Deno.run({
  cmd: ['git', 'commit', '-a', '-m', str],
  stdout: 'piped',
  stderr: 'piped',
});
const commitMsg = new TextDecoder('utf-8').decode(await res.output());
const errMsg = new TextDecoder('utf-8').decode(await res.stderrOutput());
console.log({ commitMsg, errMsg });
```

**Reply by joaomelo** [reply]
Thank you for your answer. Run something like git it not exactly my problem because Deno.run can find it as a binary, I guess. I'm talking about something like a global installed npm package.

Look:

![reproduce](https://user-images.githubusercontent.com/5727050/137894758-35e0f9fd-f688-4c93-8846-e655f63c9157.gif)

**Reply by iugo** [reply]
Maybe deno is not suitable for interactive CLI.

**Reply by joaomelo** [reply]
It is something to reflect over. Thank you @iugo.

**Comment by jtoppine** [comment]
I think what you are doing is platform dependent in a way that it makes sense for you to handle each OS manually. You don't need to create versions of your app or library tied to a platform, just add some logic to support all platforms.

I don't think automatically finding the location of binaries in various platforms is in scope for Deno built-in API. Deno API's are somewhat low level, higher level stuff like this can, and arguably should, be handled by external modules. It is something that could possibly be a useful third-party module, or maybe even std module. It is a legitimate feature request. No need to get oneself worked up on it getting moved to a discussion :)

```
if (Deno.build.os == "windows") { // "darwin" | "linux" | "windows"
  cmdprefix = ["cmd", "/c"]; // add these the wherever you call run() 
}

```
I'm a bit rusty with linux so not exactly sure what would be the equivalents in linux/darwin. Just a "/usr/bin/" + command has worked for my use cases, which are not many, though.

**Reply by joaomelo** [reply]
That seems sensible. Thank you for your support.

I take the subject as settled, but as a user, I would really enjoy if there was a cross-platform API (maybe a Deno.shell method or as a Deno.run option) to channel a command to the user's running shell.

Stay healthy.
JM

**Comment by Soremwar** [comment]
Could you try running with `allow-env`? I remember vaguely something like programs that are in global scope for your terminal need $PATH in order to do the same in Deno

**Reply by joaomelo** [reply]
I was filled with hope... but it didn't change the outcome. 
Thank you :)
